src/fd.cpp

```diff
@@ -0,0 +1,248 @@
+  }
+
+  do {
+    result = POLL_FUNC(pollfds.data(), args->num_fds, repeats ? LATER_INTERVAL : timeoutms);
+    if (args->flag->load()) goto callback;
+    if (result) break;
+  } while (infinite || (repeats-- && (timeoutms -= LATER_INTERVAL)));
+
+  if (result > 0) {
+    for (int i = 0; i < args->rfds; i++) {
+      fds[i] = pollfds[i].revents == 0 ? 0 : pollfds[i].revents & POLLIN ? 1: R_NaInt;
+    }
+    for (int i = args->rfds; i < (args->rfds + args->wfds); i++) {
+      fds[i] = pollfds[i].revents == 0 ? 0 : pollfds[i].revents & POLLOUT ? 1 : R_NaInt;
+    }
+    for (int i = args->rfds + args->wfds; i < args->num_fds; i++) {
+      fds[i] = pollfds[i].revents != 0;
+    }
+  } else if (result == 0) {
+    std::memset(fds, 0, args->num_fds * sizeof(int));
+  } else {
+    for (int i = 0; i < args->num_fds; i++) {
+      fds[i] = R_NaInt;
+    }
+  }
+
+#endif // POLLIN
+
+  callback:
+  callbackRegistryTable.scheduleCallback(later_callback, static_cast<void *>(argsptr.release()), 0, args->loop);
+
+  return 0;
+
+}
+
+// [[Rcpp::export]]
+Rcpp::RObject execLater_fd(Rcpp::Function callback, Rcpp::IntegerVector readfds, Rcpp::IntegerVector writefds,
+                                 Rcpp::IntegerVector exceptfds, Rcpp::NumericVector timeoutSecs, Rcpp::IntegerVector loop_id) {
+
+  const int rfds = static_cast<int>(readfds.size());
+  const int wfds = static_cast<int>(writefds.size());
+  const int efds = static_cast<int>(exceptfds.size());
+  const int num_fds = rfds + wfds + efds;
+  if (num_fds == 0)
+    Rcpp::stop("No file descriptors supplied");
+
+  std::shared_ptr<ThreadArgs> args = std::make_shared<ThreadArgs>();
+  std::unique_ptr<std::vector<int>> fds(new std::vector<int>());
+  fds->reserve(num_fds);
+  std::shared_ptr<std::atomic<bool>> flag = std::make_shared<std::atomic<bool>>();
+  args->flag = flag;
+  args->timeout = timeoutSecs[0];
+  args->loop = loop_id[0];
+  args->num_fds = num_fds;
+  args->rfds = rfds;
+  args->wfds = wfds;
+  args->efds = efds;
+  for (int i = 0; i < rfds; i++) {
+    fds->push_back(readfds[i]);
+  }
+  for (int i = 0; i < wfds; i++) {
+    fds->push_back(writefds[i]);
+  }
+  for (int i = 0; i < efds; i++) {
+    fds->push_back(exceptfds[i]);
+  }
+  args->fds = std::move(fds);
+
+  SEXP call = Rf_lcons(callback, Rf_cons(Rf_allocVector(LGLSXP, num_fds), R_NilValue));
+  R_PreserveObject(call);
+  args->callback = call;
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(new std::shared_ptr<ThreadArgs>(args));
+
+  tct_thrd_t thr;
+  if (tct_thrd_create(&thr, &wait_thread, static_cast<void *>(argsptr.release())) != tct_thrd_success)
+    Rcpp::stop("Thread creation failed");
+  tct_thrd_detach(thr);
+
+  Rcpp::XPtr<std::shared_ptr<std::atomic<bool>>> xptr(new std::shared_ptr<std::atomic<bool>>(flag), true);
+  SEXP body, func;
+  PROTECT(body = Rf_lang2(later_invisibleSymbol, Rf_lang2(later_fdcancel, xptr)));
+  func = R_mkClosure(R_NilValue, body, R_BaseEnv);
+  UNPROTECT(1);
```

Comment:
Could you return the xptr here? And make the closure that calls `fd_cancel`, from the R part of `later_fd`?

This kind of code (and the related code in init.c) may be easy for you to write and maintain, but it's very unnatural for me as I have always leaned very hard into Rcpp/RAII and I think the same is true for everyone on the Shiny team.

Reply:
I've implemented this in [45b665f](https://github.com/r-lib/later/pull/190/commits/45b665f91c78300ebc7cd31bc0d8389a79527356), completely mirroring what you have for `later()` itself. 

And removing the low-level code C code - although again this kind of thing is what I'm used to for pushing performance. I thought it would be useful here for our intended use cases where we'd be calling in fairly tight loops.

On the other hand, the points on maintainability are very valid and I'll be mindful in future. I'm also all for leaning into RAII 😄.